Continuous integration

mortisflux opened the pull request:
### All Submissions:

* [X] Have you followed the guidelines in our **CONTRIBUTING** document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->


<!-- These have been added according to the rOpenSci guidelines. -->
<!-- From: https://www.talater.com/open-source-templates/#/page/99 -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue. List the bug ID if applicable)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing core functionality to change) 


### New Feature Submissions checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
1. [X] Have you ensured there are no conflicts with major packages (e.g. tidyverse, MASS) or base R packages?
2. [ ] Does your new feature require documentation, and if so, have you updated the documentation?
3. [ ] Have you added tests for your features, and do your features pass all tests?
4. [ ] Have you linted your code with lintr locally prior to submission?
5. [ ] Have you styled your code with styler locally prior to submission?
6. [ ] Have you [benchmarked](https://www.alexejgossmann.com/benchmarking_r/) the performance of your code if applicable (e.g. with microbenchmark)?
7. [ ] Does your new feature comply with existing model APIs (e.g. for models, optimisation algorithms, objective functions)?


## PR description
This PR is to implement continuous integration (CI) via GitHub actions found in `.github/workflows`. The CI consists of several .yaml files, with each covering a different aspect of CI. The CI currently runs when a push/pull is made on either master or the continuous_integration branches. The continuous_integration branch commands can be deleted before merging into master.

`standard_R_ci.yaml`: is based on the actions in [r-lib](https://github.com/r-lib/actions/tree/master/examples). This action builds hydromad on Mac, Windows and Linux, using R 3.6, the current R version, and the development version (on Linux, at least). It will also:
1) Install GTK (a requirement for some of hydromad's functions)
2) Checks and installs dependencies. Note: there are some dependenices that cannot be automatically installed from CRAN because they've been archived (or weren't on CRAN in the first place) so I've provided paths to those under "Install dependencies".
3) As a part of R CMD Check, it should also run testthat on everything starting with "tests" in the `/tests` folder.

So far, the CI is failing because it is generating this warning during Install dependencies:
`‘playwith’ is not available (for R version XXX) `
which then leads to 
`##[error]Package suggested but not available: ‘playwith’`

As playwith is an optional dependency I have experimented with setting _R_CHECK_FORCE_SUGGESTS_ to a false value. Unfortunately, this does not actually solve the problem with `playwith` because by not downloading the optional dependencies, the handful functions that do require packages such as `playwith` will automatically fail their checks. 

`R_styler`: styles all .R files according to the tidyverse style (or a style of your choice, however, tidyverse is the default option). This code will automatically make a new commit with the suggested changes to the files, however, I have temporarily disabled the autocommit function (it is in the code but is commented out) because it might be a bit overwhelming to go through all the R code at the moment. 

`R_linter`: parses through all .R files and provides warnings accordingly. It does not actually modify the code itself (however, I am looking into a way to achieve this for incoming code, e.g. Alex's upcoming PR). In an ideal world we would be able to make it work similarly to the `R_styler` action, however, due to the nature of `lintr` itself this may not be possible. I'm not aware of any other R linters out there, so we may be stuck with `lintr`. Note: this action is currently failing for the same reasons as `standard_R_ci` is failing.

`C_linter`: lints all C/C++ files using [clang-format](https://github.com/DoozyX/clang-format-lint-action) and the check will fail if it finds things to lint. It makes suggestions, similarly to `R_linter`, except it is returning the linted version of the file under the original output. I think this might be more useful if it is can make autocommits in the future (e.g. using [this functionality](https://github.com/r-lib/actions/tree/master/pr-push)).


For a useful resource on better understanding workflows, please see the [official GitHub documentation](https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow).

And if you have 20 min to spare for a video crash course on GitHub workflows for R, then you can check out this [link](https://www.jimhester.com/talk/2020-rsc-github-actions/).

jnothman commented:
Nice!

There are merge conflicts with master, because this branch does not include the commits found in master, but includes some of the same content. `.DS_Store` also needs removal! ;)

> The CI currently runs when a push/pull is made on either master or the continuous_integration branches.

After merging, behaviour should be such that if any pull request is opened with `master` as the base, CI should run. Or really, any pull request.

> As playwith is an optional dependency I have experimented with setting R_CHECK_FORCE_SUGGESTS to a false value. Unfortunately, this does not actually solve the problem with playwith because by not downloading the optional dependencies, the handful functions that do require packages such as playwith will automatically fail their checks.

Standard approach consists of:
* modify the tests requiring `playwith` such that that if `playwith` is not available, the test is skipped. See `skip_if_not_installed` in testthat.
* having a CI matrix that includes `playwith` in some runs and not in others.

> I have temporarily disabled the autocommit function (it is in the code but is commented out) because it might be a bit overwhelming to go through all the R code at the moment.

Yes, keep the diff small :) But if it's easy enough, maybe we should just open a PR that runs Styler on master and see if we like it...

I've not yet worked on a repo where styling is automatically committed. One risk is that small changes by the contributor then get superseded by the automatic commit... it makes it harder to commit small changes and push if you keep needing to pull in the commit from the styler.

> R_linter: parses through all .R files and provides warnings accordingly. It does not actually modify the code itself (however, I am looking into a way to achieve this for incoming code, e.g. Alex's upcoming PR)

Not every issue identified by a linter can be automatically fixed. I would leave lintr as a pass/fail without correction. However, if lintr also fails for style issues, it needs to run *after* the Styler automatic commit.

mortisflux commented:
Thanks, Joel! Lots of good points here.

> There are merge conflicts with master, because this branch does not include the commits found in master, but includes some of the same content. `.DS_Store` also needs removal! ;)

Not again - I thought I'd added it to .gitignore on my fork!! Will remove.

> > The CI currently runs when a push/pull is made on either master or the continuous_integration branches.
> 
> After merging, behaviour should be such that if any pull request is opened with `master` as the base, CI should run. Or really, any pull request.

The setting for this should be `on: push`. I can make this change today.

> > As playwith is an optional dependency I have experimented with setting R_CHECK_FORCE_SUGGESTS to a false value. Unfortunately, this does not actually solve the problem with playwith because by not downloading the optional dependencies, the handful functions that do require packages such as playwith will automatically fail their checks.
> 
> Standard approach consists of:
> 
>     * modify the tests requiring `playwith` such that that if `playwith` is not available, the test is skipped. See `skip_if_not_installed` in testthat.
> 
>     * having a CI matrix that includes `playwith` in some runs and not in others.

That sounds like a good approach to me - I'll work on this today.

> > I have temporarily disabled the autocommit function (it is in the code but is commented out) because it might be a bit overwhelming to go through all the R code at the moment.
> 
> Yes, keep the diff small :) But if it's easy enough, maybe we should just open a PR that runs Styler on master and see if we like it...

Yes, I think this warrants a separate PR. 

> I've not yet worked on a repo where styling is automatically committed. One risk is that small changes by the contributor then get superseded by the automatic commit... it makes it harder to commit small changes and push if you keep needing to pull in the commit from the styler.

That's a good point. I can set `R_styler.yaml` to return recommended styling as script in the workflow interface rather than as a push (i.e. I can delete the script that is currently commented out if the Hydromad team agrees). 

> > R_linter: parses through all .R files and provides warnings accordingly. It does not actually modify the code itself (however, I am looking into a way to achieve this for incoming code, e.g. Alex's upcoming PR)
> 
> Not every issue identified by a linter can be automatically fixed. I would leave lintr as a pass/fail without correction. However, if lintr also fails for style issues, it needs to run _after_ the Styler automatic commit.

Ok, sure thing. In that case I'll leave `R_lintr` as is for the time being, but will look into ordering the workflow steps if we think it's best to run it after `R_styler.`

josephguillaume commented:
Thanks both! 
- I agree I would rather avoid auto-commits. Previously I've worked on repositories that used pre-commit hooks in git to do the restyling and linting. If we want to do that, it would probably be a separate PR (and linting would need to pass on the current codebase first).
- Skipping tests when a package is not available would be great.
- I assume a rebase or pull is required to fix the conflict?

mortisflux commented:
>     * I agree I would rather avoid auto-commits. Previously I've worked on repositories that used pre-commit hooks in git to do the restyling and linting. If we want to do that, it would probably be a separate PR (and linting would need to pass on the current codebase first).

Not a problem. There are no linting/styling autocommits at present and I can remove that piece of code from the .yaml file if that suits you. I think linting R code might actually be easier from within R Studio (i.e. done locally) vs linting on GitHub anyway due to the nature of the suggestions given by `lintr`. However, having lintr/styler will be useful for checking incoming code from future PRs, so I recommend we keep them in the workflow but not use the autocommit functionality - does that sound alright?

>     * Skipping tests when a package is not available would be great.

I'll work on implementing this over the next couple of days.

>     * I assume a rebase or pull is required to fix the conflict?

I think there's no need for a rebase. You can click on "Resolve conflicts" at the bottom of this page, which will then show you the conflicting files and ask you what you'd like to do. There are instructions [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-on-github) which are useful if you'd like to try to resolve them. If not, I think I can resolve them from my side.

jnothman commented:
> I think there's no need for a rebase. You can click on "Resolve conflicts" at the bottom of this page, which will then show you the conflicting files and ask you what you'd like to do. There are instructions here which are useful if you'd like to try to resolve them. If not, I think I can resolve them from my side.

Usually it's the contributor's job to clean up their diff. ;)

A merge with the master (pull this master into your working copy, then git merge that head... I can run you through this if needed; or just click "resolve conflicts" to do the same) might be easier than a rebase. A rebase requires resolving conflicts in each commit here relative to master. A merge only requires resolving conflicts between their heads.

> I think linting R code might actually be easier from within R Studio (i.e. done locally) vs linting on GitHub anyway due to the nature of the suggestions given by lintr. 

We should try to resolve many of them and submit a PR. Sometimes not all can be easily resolved, due to stylistic differences between the linter and the author. If not all can be easily resolved, you can usually disable some checks. It's useful to have the linter in the workflow once you're happy with the rules that it enforces.

josephguillaume commented:
I agree with keeping lint and styling in the CI check - thanks for putting this together!  It's just the autocommit I'd want to keep disabled.

jnothman commented on .github/workflows/standard_R_ci.yaml:
```diff
@@ -0,0 +1,102 @@
+        run: rcmdcheck::rcmdcheck(args = c("--no-manual", "--as-cran"), error_on = "warning", check_dir = "check")
+        shell: Rscript {0}
+
+      - name: Upload check results
```
does this also upload code coverage? maybe we should add that in a separate PR.

mortisflux replied:
Not at the moment, but it should be able to, perhaps with something like [covr](https://github.com/r-lib/covr). I agree, let's save this for a separate PR.

jnothman commented:
A much cleaner diff now! But the workflow is not running. Do you know why?

mortisflux commented:
> A much cleaner diff now! But the workflow is not running. Do you know why? I've never installed workflows, and have used other CI frameworks.

It is running on my fork, so I suspect it is because the workflow isn't actually within this repo yet. I think once the PR is accepted and the branched has been merged, it will run.

josephguillaume commented on .github/workflows/standard_R_ci.yaml:
```diff
@@ -0,0 +1,110 @@
+        if: runner.os == 'Windows'
+        run: |
+          choco install gtk-runtime
+
```
GTK might only be needed for playwith, in which case we'll remove this later?
Or did you see it was needed for something else?

mortisflux replied:
Yes, I believe so. We can remove this after the `playwith` dependency has been sorted out.

josephguillaume replied:
Done if you'd like to remove this prior to merging this PR.

mortisflux commented:
I've done some sleuthing into `R CMD check` and the `playwith` dependency both on GitHub and locally. Contrary to my prior thinking, it looks like setting _R_CHECK_FORCE_SUGGESTS_ to false should be sufficient for getting `R CMD check` to run the tests, because `playwith` is only used in `eventsExplorer()`, the output of which isn't examined in the test suite as far as I can see. Running `devtools::check(force_suggests = FALSE)` locally also manages to run `R CMD check` without `playwith`, so I think what I was initially seeing as an error wasn't due to `playwith`, rather, the tests themselves failing when testthat() is run.

In standard CI workflow, I've implemented:
- Running a full check first
- On the condition that the full check fails, a second check is run with _R_CHECK_FORCE_SUGGESTS_: false to skip the optional dependencies

Having this should be useful for the future if another package gets retired.

FYI, the results of the check (if it fails) are uploaded as [artifacts](https://help.github.com/en/actions/configuring-and-managing-workflows/persisting-workflow-data-using-artifacts) to GitHub that can be downloaded. The .log files in the artifacts should have the output of the errors/warnings. These warnings will also be displayed in the shell scripts on the webpage where the checks are conducted.

Other changes I've made to the workflow:
- Removed the full installation of package dependencies from `R_lintr.yaml` as we're only interested in code linting here rather than making sure the package builds, so this shouldn't cause a problem.
- Minor name changes to step names for the sake of consistency (e.g. "GTK Windows" to "Install GTK for Windows") in `standard_ci.yaml`
- Changed the build matrix in `standard_ci.yaml` such that instead of R "4.0" we have "release". This is a bit of future proofing so that 4.0 isn't hardcoded as the current version.

The conflicts have also been resolved and there is no autocommit of code styling, so I think this is ready for merging!

jnothman commented:
> On the condition that the full check fails, a second check is run with R_CHECK_FORCE_SUGGESTS: false to skip the optional dependencies

It can be useful in all cases to run the tests without optional dependencies installed, as this ensures that they do indeed remain optional.

> The conflicts have also been resolved and there is no autocommit of code styling, so I think this is ready for merging!

I remain surprised that we can't see the output of the checks run on this pull request, and have to merge first.

josephguillaume commented:
The checks are visible on the source repository:
https://github.com/Sydney-Informatics-Hub/hydromad/tree/continuous_integration

jnothman commented on .github/workflows/R_lintr.yaml:
```diff
@@ -29,5 +29,5 @@ jobs:
       - name: Lint
-        run: lintr::lint_package()
+        run: lintr::lint_package(linters = lintr::with_defaults(object_name_linter = NULL))
         shell: Rscript {0}
```
Why is this not 
```suggestion
        shell: Rscript -e {0}
```
?

jnothman commented on .github/workflows/R_lintr.yaml:
```diff
@@ -0,0 +1,33 @@
+        shell: Rscript {0}
+
+      - name: Lint
+        run: lintr::lint_package(linters = lintr::with_defaults(object_name_linter = NULL))
```
Maybe if we do something like:
```suggestion
        run: |
          results = lintr::lint_package(linters = lintr::with_defaults(object_name_linter = NULL))
          if (length(results)) { print(results); quit(status=1) }
```
we will get the failing exit status and if we're lucky, a proper printout

mortisflux replied:
I'll give this a go, thanks Joel!

mortisflux replied:
The R linter is now coming up with a true negative alert! :)

mortisflux replied:
However, I don't think this is due to the linting itself ... there is an error popping up:
```
##[error]Error: unexpected '/' in "/"
Execution halted
##[error]Process completed with exit code 1.
```

jnothman replied:
So, a false true negative?? :D

jnothman replied:
It seems Rscript doesn't like receiving two commands in one.

This a more R-like single statement, and might make it happy:
```
if ( length(lintr::lint_package(linters = lintr::with_defaults(object_name_linter = NULL)) -> results)) { print(results); quit(status=1) }
```

jnothman replied:
I've just pushed that change to this branch. I had mean to submit it to you for review first, @mortisflux, but it seems my fingers went into automatic.

mortisflux replied:
The lint failed (for the correct reasons). No nice printout of the linting as in R Studio though ...

jnothman commented on .github/workflows/C_linter.yaml:
```diff
@@ -0,0 +1,24 @@
+on:
```
how about we run on all pushes and pull requests. why not?

mortisflux replied:
Re: push pull requests - you mean regardless of the branch? We can do that, as it'll be useful if anyone wants to develop on another branch and see the output of the workflows.

josephguillaume commented:
Fantastic! You two are awesome. Reading the discussion was quite a rollercoaster.
Let me know when you're ok for me to merge.

jnothman commented:
> Your tweak worked!! It's nicely printing the output now!

Ugly, though.

And I still don't get how the annotations were created in the original example, except that you need a "problem matcher"... r-lib has problemMatchers, but I don't really see how they apply here.

To match something like:
`inst/doc/hydromad_results.R:9:24: style: Place a space before left parenthesis, except in a function call.`
I think we could now add something like:
```json
{
    "problemMatcher": [
        {
            "owner": "R_lintr",
            "pattern": [
                {
                    "regexp": "^(.+):(\\d+):(\\d+):\\s([a-z]+):\\s(.+)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "severity": 4,
                    "message": 5
                }
            ]
        }
    ]
}
```

jnothman commented:
> Fantastic! You two are awesome. Reading the discussion was quite a rollercoaster.
Let me know when you're ok for me to merge.

Merge can happen when you're ready. I'm trying to make annotations work at https://github.com/Sydney-Informatics-Hub/hydromad/runs/806369228?check_suite_focus=true. But because we have so many lint outputs, the annotations won't work properly anyway; they're capped at 50 :)

jnothman commented:
Annotations not working for me :( Oh well... maybe later.

jnothman commented:
I now understand that converting the output to annotations is what obscures the output in the log. Shrug. I'm reverting to the version which quits correctly produces "Annotations" but which makes the log less useful. Hopefully, we will find these annotations useful once style issues are fixed.
![image](https://user-images.githubusercontent.com/78827/85701770-160d5e00-b721-11ea-9cbf-e6adbcab4f80.png)

josephguillaume commented:
It turns out that github makes it difficult to copy the clinter patch from the browser (only selects small sections at once) and removing the timestamps to obtain a non-corrupt patch is turning out more difficult than expected. How difficult would it be to make the patch available as an artifact?

josephguillaume commented:
Also, it turns out rcpp regenerates rcpp_exports.cpp on every build, so we'd probably want to exclude it from the linting?

josephguillaume commented:
There's a strange error for standard_ci:
> unable to access index for repository https://cloud.r-project.org/bin/macosx/contrib/4.1

Current R version is 4.0.2. R 4.1 doesn't exist yet, so I'm not sure why it's looking for that repository.

josephguillaume commented:
It seems all of the standard_ci annotations are listed as line 1 in .github, which limits their usefulness. If there isn't an easy fix, let's open an issue to address this down the track.

jnothman commented:
> Also, it turns out rcpp regenerates rcpp_exports.cpp on every build, so we'd probably want to exclude it from the linting? Or maybe I should just remove it from the repo.

If it's going to be regenerated, removing it and adding to `.gitignore` seems appropriate.

josephguillaume commented:
> > Also, it turns out rcpp regenerates rcpp_exports.cpp on every build, so we'd probably want to exclude it from the linting? Or maybe I should just remove it from the repo.
> 
> If it's going to be regenerated, removing it and adding to `.gitignore` seems appropriate.

Sounds good. Can do.

mortisflux commented:
> > How difficult would it be to make the patch available as an artifact?
> 
> Seems like github workflows doesn't make this as easy as hoped - let me know if you'd rather leave this to a later date...

It's being very pesky! I've tried to force the generation of an artifact with the previous workflow as well as try to run a shell script using clang-format but it hasn't worked thus far.

I'm going to revert the changes back and suggest we try again at a later date. To do C linting meantime, I would advise downloading the action [logs](https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#downloading-logs) and then editing out the the timestamps. Sublime Text (and Notepad ++ too) has a neat function where you [split your file into lines](https://stackoverflow.com/questions/39556514/sublime-text-3-how-to-edit-multiple-lines) and then can edit across all lines ... that's a quick-ish way to get rid of those timestamps.

josephguillaume commented:
Let's merge and add issues for the last pesky details?

mortisflux commented:
> Let's merge and add issues for the last pesky details?

Sounds good to me! 

I also had a look into the potential problems with the R-devel version ... I think it's working ok, not sure why that error is popping up. The R-devel version definitely installs with gfortran as per the setup-r@master step:

` sudo installer -pkg /Users/runner/runners/2.263.0/work/_temp/R-devel.pkg -target /. 
installer: Package name is R 4.1.0 for macOS. 
installer: The install was successful.`

and there's a line under R CMD Check in the logs that states:

`* using R Under development (unstable) (2020-06-21 r78727)`

josephguillaume commented:
Done. Thanks!
I've added issues: https://github.com/josephguillaume/hydromad/issues/81 https://github.com/josephguillaume/hydromad/issues/82 https://github.com/josephguillaume/hydromad/issues/83 and removed RcppExports

Maybe check if I missed any other issues, but otherwise I think the next priority is a PR running styler?
